docs/RFCS/20191028_easier_hash_sharded_indexes.md

```diff
@@ -0,0 +1,220 @@
+    device_id
+        UUID,
+    shard
+        INT8
+        AS (fnv32(device_id) % 8) STORED
+        CHECK (shard IN (0, 1, 2, 3, 4, 5, 6, 7)),
+    event_id
+        UUID,
+    ts
+        TIMESTAMP,
+    data
+        JSONB,
+    PRIMARY KEY (device_id, ts, event_id),
+    INDEX (shard, ts)
+);
+```
+
+This isn't too big of a lift here because the device ID is easy to hash in sql. Imagine
+instead we had a primary key based on some other features:
+
+```sql
+CREATE TABLE events (
+    product_id
+        INT8,
+    owner
+        UUID,
+    serial_number
+        VARCHAR,
+    event_id
+        UUID,
+    ts
+        TIMESTAMP,
+    data
+        JSONB,
+    PRIMARY KEY (product_id, owner, serial_number, ts, event_id),
+    INDEX (ts)
+);
+```
+
+In order to shard this we'll need something like:
+
+```sql
+CREATE TABLE events (
+    product_id
+        INT8,
+    owner
+        UUID,
+    serial_number
+        VARCHAR,
+    shard
+        INT8
+        AS (
+            fnv32(
+                concat(hex(product_id)),
+                owner::STRING,
+                serial_number
+            )
+            % 8
+        ) STORED
+        CHECK (shard IN (0, 1, 2, 3, 4, 5, 6, 7)),
+    event_id
+        UUID,
+    ts
+        TIMESTAMP,
+    data
+        JSONB,
+    PRIMARY KEY (
+        product_id,
+        owner,
+        serial_number,
+        ts,
+        event_id
+    ),
+    INDEX (shard, ts)
+);
+```
+
+We can see that this is starting to get heavy. The proposal is that we shoulder the burden
+of hashing and installing a check constraint behind a new syntax:
+
+```sql
+CREATE TABLE events (
+    product_id
+        INT8,
+    owner
+        UUID,
+    serial_number
+        VARCHAR,
+    event_id
+        UUID,
+    ts
+        TIMESTAMP,
+    data
+        JSONB,
+    PRIMARY KEY (product_id, owner, serial_number, ts, event_id),
+    INDEX (ts) USING HASH (product_id, owner, serial_number) WITH (BUCKET_COUNT=8)
+);
+```
+
+# Benchmarks 
+
+ We benchmarked the sequential kv workload against two schemas:
+
+ (1) Using hash sharded primary index on `k`.
+```sql
+CREATE TABLE kv (
+    k INT8 NOT NULL,
+    v bytes NOT NULL,
+    shard 
+        INT8 AS (k % 8) stored 
+        CHECK (shard IN (0, 1, 2, 3, 4, 5, 6, 7)), 
+    PRIMARY KEY (shard, k)
+);
+```
+
+(2) Using primary index on `k`.
+```sql
+CREATE TABLE kv (
+    k INT8 NOT NULL,
+    v BYTES NOT NULL,
+    PRIMARY KEY (k)
+);
+```
+
+ The following benchmarks were all run on a 7 node GCE cluster with `n1-standard-32`
+ machines. We first demonstrate write performance with `kv --sequential --duration=5m` at
+ varying levels of concurrency (256, 384, 512). Note that this is a write-only workload.
+
+
+![Latency](https://user-images.githubusercontent.com/10788754/67722397-63c5a100-f9af-11e9-8316-a173e52fe20b.png)
+
+
+![Throughput](https://user-images.githubusercontent.com/10788754/67733735-8ec2eb80-f9d5-11e9-889f-38b4466d7234.png)
+
+## Drawbacks / Future work
```

Comment:
We spoke offline but talk through what it takes to plumb the calculation of shard through to the select. 

The idea is that for these point lookups, when this project is actually done, we won't have this inefficiency.